### PR TITLE
tests: kernel: timer: Fix failing tests for custom k_busy_wait()

### PR DESCRIPTION
**Pull request by adamkondraciuk**
This commit provides an additional threshold value for comparison when a custom k_busy_wait() implementation is chosen.

### CODE REVIEW

File: tests/kernel/timer/timer_api/src/main.c [hunk 1]
```diff
@@ -28,15 +28,26 @@ struct timer_data {
...
 #if CONFIG_NRF_RTC_TIMER
 /* On Nordic SOCs one or both of the tick and busy-wait clocks may
  * derive from sources that have slews that sum to +/- 13%.
  */
 #define BUSY_TICK_SLEW_PPM 130000U
 #else
-/* On other platforms assume the clocks are perfectly aligned. */
-#define BUSY_TICK_SLEW_PPM 0U
+/* On other platforms the maximum skew allowed is +/- 2%. */
+#define BUSY_TICK_SLEW_PPM 20000U
```
Comment (andyross) [inline tests/kernel/timer/timer_api/src/main.c:43]: Would prefer to make excuses based on specific platform clock behavior.  A busy_wait being "custom" doesn't necessarily mean it's inaccurate.    See the comment above about nRF, where the skew is an understood thing that can't be reasonably worked around (well, without changing the CPU clock to be a PLL based off the same crystal the timer driver uses vs. the internal oscillator, which AIUI has power impact that Nordic doesn't want to accept on the default config).

Comment (adamkondraciuk) [reply tests/kernel/timer/timer_api/src/main.c:42]: > Would prefer to make excuses based on specific platform clock behavior. A busy_wait being "custom" doesn't necessarily mean it's inaccurate. See the comment above about nRF, where the skew is an understood thing that can't be reasonably worked around (well, without changing the CPU clock to be a PLL based off the same crystal the timer driver uses vs. the internal oscillator, which AIUI has power impact that Nordic doesn't want to accept on the default config).

Okay, how about that?

**Review by andyross — DISMISSED** [review]
Thanks

**Review by TaiJuWu — DISMISSED** [review]

File: tests/kernel/timer/timer_api/testcase.yaml [hunk 2]
```diff
@@ -4,6 +4,8 @@ tests:
       - kernel
       - timer
       - userspace
+    platform_exclude:
+      - nrf54h20dk/nrf54h20/cpuppr
```
Comment (andyross) [inline tests/kernel/timer/timer_api/testcase.yaml:8]: That's frustrating.  Can you file a bug on this, we should probably look at splitting this into two tests then.  timer_api is one of the core/most-important/best-validation-smokes tests and we really don't want to be disabling it for technicalities.

Comment (andyross) [reply tests/kernel/timer/timer_api/testcase.yaml:8]: Also, and I swear I'm not pointing fingers here, failures in timer_api are often very subtle things that look like "the test is wrong" or "it's just random/unfixable".  It's happened again and again that tiny edge case failures on obscure platforms or configurations turn out to be evidence of bugs in the driver or timer subsystem.  We really want this test to be on for everyone.

**Comment by MaureenHelm** [comment]
@andyross please revisit

**Review by kl-cruz — APPROVED** [review]

**Review by cfriedt — CHANGES_REQUESTED** [review]
Just wanted to double-check usage of the term "slew", which is typically used in analog circuits (e.g. opamps), and describes dv/dt at a single test point.

https://en.wikipedia.org/wiki/Slew_rate

"Skew" is a phenomenon that appears due to the propagation delay of a clock signal measured at one point vs at another point and also differences in rates between independent clock signals.

https://en.wikipedia.org/wiki/Clock_skew

Apologies if this was already brought up. English can be a bit of a silly language sometimes.

**Comment by adamkondraciuk** [comment]
> Just wanted to double-check usage of the term "slew", which is typically used in analog circuits (e.g. opamps), and describes dv/dt at a single test point.
> 
> https://en.wikipedia.org/wiki/Slew_rate
> 
> "Skew" is a phenomenon that appears due to the propagation delay of a clock signal measured at one point vs at another point and also differences in rates between independent clock signals.
> 
> https://en.wikipedia.org/wiki/Clock_skew
> 
> Apologies if this was already brought up. English can be a bit of a silly language sometimes.
Hmmm... seems that neither "slew" nor "skew" is a good term here. The skew is an effect of the propagation of one clock source while in our case the thing is about deviation between two independent clock sources (both - RTC and GRTC are clocked by 32k oscillator while in our case 'k_busy_wait()' is clocked by the core clock) ... Am I right?

**Comment by cfriedt** [comment]
> Hmmm... seems that neither "slew" nor "skew" is a good term here. The skew is an effect of the propagation of one clock source while in our case the thing is about deviation between two independent clock sources (both - RTC and GRTC are clocked by 32k oscillator while in our case 'k_busy_wait()' is clocked by the core clock) ... Am I right?

Yeah, I think you're right. Deviation is probably the more appropriate term. FWIR in rfsoc's and e.g. in bluetooth, the differences between two clocks would be specified as a deviation in terms of ± some number of PPM.

I'll just ping all of the time lords I can think of 
@cvinayak @npitre @fgrandel @andyross - does the above sound accurate?

**Comment by npitre** [comment]
+1 for "deviation"

**Comment by cvinayak** [comment]
Thinking aloud here, with opportunity to learn too; a clock source has _accuracy_ in ppm, that leads to _jitter_ in a measured interval, over time of multiple such periodic interval a probable _drift_ in relation to a perfect/accurate clock source. Now, with two clock sources of differing _accuracy_ in ppm, the _drift_ will be sum of their _accuracy_ over the interval.

From the nRF52 product specification, the High Frequency Clock (internal RC typically unless External Crystal use is requested) that drives the CPU is typical +/- 1.5% at 64 MHz(now I do not know how to convert this to ppm! maybe I should learn it), and this will be the `k_busy_wait()` clock source. The Low Frequency Clock (typically 32 KHz crystal) at best be minimum of 0 to 20 ppm is driving the NRF_RTC peripheral, that is used as system timer. And I have experienced practically these crystals give ~100 ppm accuracy with all the software related jitter.

Trying to understand what this test is checking, it looks like it is comparing the _absolute_ value of the `delta_ticks`; hence, comparing with _deviation_ sound right to me.

+1 for _deviation_

**Comment by fgrandel** [comment]
> Hmmm... seems that neither "slew" nor "skew" is a good term here. The skew is an effect of the propagation of one clock source while in our case the thing is about deviation between two independent clock sources

Interesting discussion. :-)

**TLDR;**

Time-related terms are well standardized so while all proposals made so far are certainly not wrong I propose that we stick to terms frequently used in standards.

IMO the most appropriate standard nomenclature would be **+/-x% frequency offset of the busy timer oscillator as compared to the kernel timer oscillator.**

* You can replace _oscillator_ by _tick source_ or leave it out entirely.
* You can replace _frequency offset_ by _frequency deviation_ although this is used less commonly in standards.
* You should say clearly that it is _frequency_ offset that is measured, not jitter or (frequency) drift or absolut time deviation.
* You should _not_ call those timers a _clock source_. They measure relative time distances or frequencies or at most arbitrary uptime. A proper clock source would also have to provide a time scale and time origin. Note: Linux's sloppy use of the word "clock source" may be justifiable because they always add time origin/scale downstream to their clock sources, which we don't in general to our kernel/busy timers.

**Rationale:**

Because I love time measurement I took the opportunity to actually consult a few standards, but of course this goes completely overboard wrt what we need here - just for those interested...

Common standards are (among others):
- IEEE 1139-2022 (hidden behind a pay wall)
- [MIL-PRF-55310](https://nepp.nasa.gov/DocUploads/1F3275A6-9140-4C0C-864542DBF16EB1CC/MIL-PRF-55310.pdf)
- [NIST technical note 1337](https://tf.nist.gov/general/pdf/868.pdf)
- [ITU-T G.801](file:///home/fgrandel/Insync/fgrandel@gmail.com/Development/TSN/standards/T-REC-G.810-199608-I!!PDF-E.pdf)

> a clock source has accuracy

That is correct except that we do not necessarily look at a clock source here, see above.

**Accuracy** can either mean "trueness" (i.e. closeness to a reference) as opposed to "precision" (i.e. measurement granularity) or it can mean a combination of "trueness" and "precision" (ISO-5725-1).

Neither trueness nor precision (in that well-defined sense) seem to be the problem here, though. Even if both clock sources (probably oscillators) were highly inaccurate they'd not necessarily "deviate" or "drift" relatively to each other if they deviate in the same way. This is why the test doesn't fail on a single clock source even if that clock source was highly inaccurate.

> now I do not know how to convert this to ppm

1 % = 10 ‰ = 10000 ppm
1ppm = 1000 ppb

> that leads to jitter in a measured interval

Not really. Random or bounded **jitter** or frequency stability leads per definition to constant accuracy loss over time and is therefore mostly independent of the measurement interval unless it is very short such that its randomness matters.

NIST definition: "short term phase variation of the significant instants of a digital signal from their ideal positions in time".

Jitter is just one of several sources of oscillator error. It's probably also not the one that is most significant here, as it has constant or even (in the symmetric case) zero accurracy effect in the long term by definition. It also has a bounded effect on average precision.

Other common sources of oscillator error are frequency offset (~ "time drift") and frequency drift (~ "time drift of second order").

**Frequency offset** (or frequency accuracy or frequency deviation) over time can be measured absolutely (e.g. in Hertz) or relatively (e.g. in %, ppm, ppb, ...) as compared to another frequency source (our case) or as compared to nominal frequency (which is what you usually find in data sheets). When we colloquially speak of clock or oscillator "drift" or "accuracy", what we usually mean is frequency offset as compared to the reference frequency. This is also the measure you cited from the data sheet.

**Frequency drift** over time can be measured in "offset per time unit", i.e. it measures Hertz/ppm/... per second/minute/hour/year or sometimes other parameters like temperature, etc. It is the "slope" of frequency over time. Sometimes data sheets characterize frequency drift due to temperature, radiation or ageing. Secondary parameters like these can be measured over time, too, so that the overall drift in an oscillator "quality" measurement can usually given per time interval (e.g. per one timer period).

Eliminating frequency offset and drift between two oscillators (or other tick sources) is called **syntonization** as opposed to **synchronization** of clock sources which additionally requires negotiation of a common time origin and time scale.

The three most common sources of oscillator error can be approximated in the following time-domain formulae that calculate the reference timepoint (as measured by a reference clock with *f_ref*) of the t'th oscillator "tick" x(t):

    x(t) = x(0) + y*t + (D/2) * t^2 + Phy(t) / (2*pi*f_ref)

Where (simplified):
x(t): the time of the t'th tick as measured by the reference clock
x(0): the tick that started measurement as expressed in the reference clock's time scale
y: frequency offset (linear phase error)
D: frequency drift (quadratic phase error)
Phy(t): jitter or phase fluctuation i.e. the inverse of frequency stability (random phase error)
f_ref: the nominal (reference) frequency

Or alternatively and more precisely:

    x(t) = x(0) + tau * sum_0...t-1[y(t)]

Where (simplified):
y(t): averaged normalized frequency offset (aka average fractional frequency deviation) between tick t and tick t+1 (a value of 1 means no offset)
tau: nominal tick period

Of course lots of other formulae in the time and frequency domain have been proposed to characterize oszillators.

Based on such a time series of "ticks", the "quality" of an oscillator can be characterized by lots of aggregate measures, e.g. ADEV: the Allan deviation, MDEV: the modiﬁed Allan deviation, TDEV: the time deviation, TIErms : the root mean square of the time interval error (TIE), and MTIE: the maximum time interval error, etc. This only scratches the surface. Look at some of the above cited standards to get an idea how complex actual oscillator characterization can be.

With all this in mind what actually seems to matter most here is what is usually called "frequency offset". Frequency drift should be less important due to the short measurement interval. Jitter should be largely irrelevant as it should explain a very small fraction of the overall error only.

File: tests/kernel/timer/timer_api/Kconfig [hunk 3]
```diff
@@ -4,4 +4,7 @@
 config SYS_CLOCK_TICKS_PER_SEC
 	default 16384 if NRF_RTC_TIMER && TICKLESS_KERNEL
 
+config ASSERT_VERBOSE
```
Comment (fgrandel) [inline tests/kernel/timer/timer_api/Kconfig:7]: As Kconfig variables are in the global namespace, this name is too generic IMO. You should prefix it with something that limits it to the timer API namespace.
